x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+export type DimensionRole =
+  | 'splitChart'
+  | 'series'
+  | 'primary'
+  | 'secondary'
+  | 'color'
+  | 'size'
+  | string; // Some visualizations will use custom names that have other meaning
+
+export interface TableMetaInfo {
+  columnId: string;
+  operation: Operation;
+}
+
+export interface DatasourceSuggestion<T> {
+  state: T;
+  tableMetas: TableMetaInfo[];
+}
+
+/**
+ * Interface for the datasource registry
+ */
+export interface Datasource<T> {
+  // For initializing from saved object
+  // initialize: (state?: T) => Promise<T>;
+
+  renderDataPanel: (props: DatasourceDataPanelProps) => void;
+
+  toExpression: (state: T) => string;
+
+  getDatasourceSuggestionsForField: (state: T) => Array<DatasourceSuggestion<T>>;
+  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;
+
+  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
+  // publicAPI: {
+  //   [key in keyof DatasourcePublicAPI]: (
+  //     state: T,
+  //     setState: (newState: T) => void
+  //   ) => DatasourcePublicAPI[key]
+  // };
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
+    props: DatasourceDimensionPanelProps
+  ) => (domElement: Element, operations: Operation[]) => void;
+
+  removeColumnInTableSpec: (columnId: string) => TableSpec;
+  moveColumnTo: (columnId: string, targetIndex: number) => void;
+  duplicateColumn: (columnId: string) => TableSpec;
+}
+
+export interface DatasourceDataPanelProps {
+  domElement: Element;
+}
+
+// The only way a visualization has to restrict the query building
+export interface DatasourceDimensionPanelProps {
+  // If no columnId is passed, it will render as empty
+  columnId?: string;
+
+  // Visualizations can restrict operations based on their own rules
+  filterOperations: (operation: Operation) => boolean;
+
+  // Visualizations can hint at the role this dimension would play, which
+  // affects the ordering of the query
+  suggestedPriority?: DimensionPriority;
+}
+
+export interface DatasourceValidationError {
+  type: 'error';
+}
+
+export interface DatasourceValidationValid {
+  type: 'valid';
+}
+
+export type DataType = 'string' | 'number' | 'date' | 'boolean';
+
+// An operation does not represent a column in the datatable
+export interface Operation {
+  // Operation ID is a reference to the operation
+  id: string;
+  // User-facing label for the operation
+  label: string;
+  dataType: DataType;
+  // A bucketed operation has many values the same
+  isBucketed: boolean;
+
+  // Extra meta-information like cardinality, color
+}
+
+export interface TableSpecColumn {
+  // Column IDs are the keys for internal state in data sources and visualizations
+  columnId: string;
```

Comment:
I'm wondering if this is where `suggestedPriority` belongs, since that is a concept that needs to be transferred between the datasource and the visualization. In fact, I'm a bit fuzzy on where transfer of query-like information happens between visualizations and datasources.

Reply:
There is not shared ownership of the TableSpec. The visualization can use `suggestedPriority` on a dimension panel, but it should just affect the default ordering. All re-ordering will be handled via dimension panel.